### ISSUE TITLE
Create output directories and build file registry based on `MODEL_SPEC`

### ISSUE THREAD

**Issue by emlys** [natcap/invest#1378]
All models do some setup at the beginning of `execute` that includes creating the output directory at `args['workspace_dir']` and creating any sub-directories in it. Many models also use `utils.build_file_registry`, which takes in some data that is redundant to the `MODEL_SPEC`. This setup could be done automatically given the information in `MODEL_SPEC['outputs']`.

**Comment by davemfish** [comment]
>So we'd already need to rethink HQ's logic if we wanted to use build_file_registry and remove all custom filepath creation / mapping.
With that in mind, I suppose the bigger question is: Should we be aiming for a solution that we can apply across all models as they currently exist, with as little refactoring required as possible? Or should we be aiming for a more "ideal" solution that works across the models currently using build_file_registry, and then perhaps aim to bring other models in line with this standardization at some point in the future?

I think these are good questions. In my opinion the purpose of `build_file_registry` is just to be convenient for the developer. And that convenience mainly comes up when someone is writing a new invest model. In that case, it does sound convenient to be able to pass the `MODEL_SPEC` into `build_file_registry`. But if it's inconvenient to make the spec digestible by that function, then it loses the benefit. I don't think we should go out of our way to use the function if it's not convenient to do so.

**Comment by emlys** [comment]
For the seasonal water yield example, we could use the simple, verbose approach of instantiating a separate output object for each file:
```
spec.SingleBandRasterOutput(
      id=f'qf_{month}',
      path=f'intermediate/qf_{month}.tif',
      about=f'Map of quickflow values for month {month}',
      data_type=float,
      units=u.millimeter
) for month in range(1, 13)
```
This doesn't take up any extra space in the code, and it's easier to understand. The downside is it's overly verbose for some cases where we're _using_ the spec, such as documentation. But I'd argue that it's better to be too verbose rather than unclear.

**Comment by emlys** [comment]
I agree with @davemfish that we don't need to force this solution to address every single case. But if we want to solve the cases where file names contain a variable that can take arbitrary values, a different data structure might help. Something like a `FileRegistry` class could track file name patterns and handle formatting them given a variable.

**Comment by megannissel** [comment]
> For the seasonal water yield example, we could use the simple, verbose approach of instantiating a separate output object for each file

I like this a lot, honestly. I think I got caught up over-engineering a general solution that could potentially address _every_ case, but this is so much cleaner / clearer for cases like this.

**Comment by megannissel** [comment]
@emlys I think this issue was fully addressed by your implementation of the `FileRegistry` in #2131 and some subsequent updates. Does anything here feel outstanding to you, or can we close this one?

**Comment by emlys** [comment]
I think this can be closed! Thanks for flagging it.
